### PR TITLE
Add documentation on the `cx16` CPU capabiilty that we require

### PR DESCRIPTION
**Pull request by staticfloat**
VirtualBox fails to run the typical Julia binaries because we require `cx16` but don't say that anywhere outside of `processor_x86.cpp`, so I added it to the docs.

### CODE REVIEW

**Comment by Keno** [comment]
Maybe explicitly check for this in the arch detection code and print an appropriate error and a reference to the documentation?

File: doc/src/devdocs/sysimg.md
Comment (nalimilan) [inline doc/src/devdocs/sysimg.md]: Lowercase "m".

File: doc/src/devdocs/sysimg.md
Comment (vchuravy) [inline doc/src/devdocs/sysimg.md]: @yuyichao Do you recall why we require `cx16`

Comment (yuyichao) [reply doc/src/devdocs/sysimg.md]: Yes, I believe this is the still complicated issue of ABI compliance of 128bit atomics.

1. GCC behavior. It always emit a `__atomic_*` libcall.
2. Clang behavior. It uses `cmpxchg16b` instruction when available and `__sync_lock_test_and_set_16` if not.

There are arguments suggesting that the GCC one is the only correct behavior since the clang behavior does not allow atomic load on constant memory but it does mimic the current implementation in `libatomic` and it seems that they are not changing that any time soon...

For us, if we don't require `cx16`, we need a way to inject a `__sync_lock_test_and_set_16` implementation that makes `llvm` happy.

**Review by yuyichao — APPROVED** [review]
Seems fine. `-cx16` is allowed for this reason as said in the comment though I don't remember if I ever actually tested it.... Not really documented anywhere else likely because I'm not really sure who will need it.......... Glad it works I guess.

If systems that doesn't have `cx16` is actually out there then the 128bit atomic support should be fixed so that we don't require this feature. But this should be good enough for now.

File: doc/src/devdocs/sysimg.md
Comment (vchuravy) [reply doc/src/devdocs/sysimg.md]: Right so we would have #18706

**Comment by staticfloat** [comment]
This is confirmed to work as expected.  When running a typical binary on a platform that does not support the `cx16` instruction, you get the following:

```
ERROR: Your CPU does not support the CX16 instruction, which is required by this version of Julia!  This is often due to running inside of a virtualized environment.  Please read https://docs.julialang.org/en/stable/devdocs/sysimg/ for more.
```

However, if the host platform supports the `cx16` instruction, or the sysimg explicitly contains a target that does not require the `cx16` instruction, that warning will not be displayed.

**Comment by staticfloat** [comment]
Can I get a review approval for the C changes added on after Yichao's approval?

File: src/processor_x86.cpp
Comment (yuyichao) [inline src/processor_x86.cpp]: Maybe set this to `true` on 32bits (just replace `false` with `sizeof(void*) == 4`) so that the edge case of user explicitly specifying `+cx16` on 32bit will always be ignored. There's little reason for the user to specify that though so not a major issue...

**Review by yuyichao — APPROVED** [review]

**Comment by staticfloat** [comment]
CI errors unrelated.

**Pull request merged**
